toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -358,6 +398,39 @@ class ErrorTrackingDiagnosticConsumer : public DiagnosticConsumer {
   bool seen_error_ = false;
 };
 
+// An RAII object that denotes a scope in which any diagnostic produced should
+// be annotated in some way.
+//
+// This object is given a function `annotate` that will be called with a
+// `DiagnosticBuilder& builder` for any diagnostic that is emitted through the
+// given emitter. That function can annotate the diagnostic by calling
+// `builder.Note` to add notes.
+template <typename LocationT, typename AnnotateFn>
+class DiagnosticAnnotationScope
+    : private DiagnosticEmitter<LocationT>::DiagnosticAnnotationScopeBase {
+  using Base =
+      typename DiagnosticEmitter<LocationT>::DiagnosticAnnotationScopeBase;
+
+ public:
+  DiagnosticAnnotationScope(DiagnosticEmitter<LocationT>* emitter,
+                            AnnotateFn annotate)
+      : Base(emitter), annotate_(annotate) {}
+
+ private:
+  auto Annotate(
+      typename DiagnosticEmitter<LocationT>::DiagnosticBuilder& builder)
+      -> void override {
+    annotate_(builder);
+  }
+
+  AnnotateFn annotate_;
+};
+
+template <typename LocationT, typename AnnotateFn>
+DiagnosticAnnotationScope(DiagnosticEmitter<LocationT>* emitter,
```

Comment:
Am I missing something here, or is this undefined / unused?

Reply:
This is a [deduction guide](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction).